addons/web_editor/static/src/components/media_dialog/image_selector.js

```diff
@@ -0,0 +1,256 @@
+/** @odoo-module */
+
+import { useService } from '@web/core/utils/hooks';
+import { getCSSVariableValue } from 'web_editor.utils';
+import { Attachment, FileSelector, IMAGE_MIMETYPES, IMAGE_EXTENSIONS } from './file_selector';
+
+const { useRef, useState, useEffect } = owl;
+
+export class AutoResizeImage extends Attachment {
+    setup() {
+        super.setup();
+
+        this.image = useRef('auto-resize-image');
+        this.container = useRef('auto-resize-image-container');
+
+        this.state = useState({
+            loaded: false,
+        });
+
+        useEffect(() => {
+            this.image.el.addEventListener('load', () => this.onImageLoaded());
+            return this.image.el.removeEventListener('load', () => this.onImageLoaded());
+        }, () => []);
+    }
+
+    onImageLoaded() {
+        const aspectRatio = this.image.el.offsetWidth / this.image.el.offsetHeight;
+        const width = aspectRatio * this.props.minRowHeight;
+        this.container.el.style.flexGrow = width;
+        this.container.el.style.flexBasis = `${width}px`;
+        this.state.loaded = true;
+    }
+}
+AutoResizeImage.template = 'web_editor.AutoResizeImage';
+
+export class ImageSelector extends FileSelector {
+    setup() {
+        super.setup();
+
+        this.rpc = useService('rpc');
+
+        this.state.libraryMedia = [];
+        this.state.libraryResults = null;
+        this.state.isFetchingLibrary = false;
+        this.state.searchService = 'all';
+        this.state.showOptimized = false;
+
+        this.uploadText = this.env._t("Upload an image");
+        this.urlPlaceholder = "https://www.odoo.com/logo.png";
+        this.addText = this.env._t("Add URL");
+        this.searchPlaceholder = this.env._t("Search an image");
+        this.urlWarningTitle = this.env._t("Uploaded image's format is not supported. Try with: " + IMAGE_EXTENSIONS.join(', '));
+        this.allLoadedText = this.env._t("All images have been loaded");
+        this.showOptimizedOption = this.env.debug;
+        this.MIN_ROW_HEIGHT = 128;
+
+        this.fileMimetypes = IMAGE_MIMETYPES.join(',');
+
+        useEffect(() => {
+            const initWithMedia = async () => {
+                if (this.props.media && this.props.media.tagName === 'IMG') {
+                    let selectedMedia = this.state.attachments.filter(attachment => {
+                        if (this.props.media.dataset.originalSrc) {
+                            return this.props.media.dataset.originalSrc === attachment.image_src;
+                        }
+                        return this.props.media.getAttribute('src') === attachment.image_src;
+                    })[0];
+                    if (selectedMedia) {
+                        await this.selectAttachment(selectedMedia, false);
+                    }
+                }
+            };
+
+            initWithMedia();
+        }, () => []);
```

Comment:
- I used useEffect instead of onMounted/onWillUnmount because useEffect can do it all (in the event of those hooks being removed) but I agree that their names hooks are clearer. I am changing to that.
selectAttachment is async, because it will also save in some cases, but this is probably a bad design, and save could be called independently from selectAttachment
- I think the async calls (to the services async functions) are protected by using the hook useService